File: 141-linked-list-cycle/141-linked-list-cycle.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} head
# @return {Boolean}
def hasCycle(head)
    w = head
    r = head
    
    while r and r.next
        w = w.next
        r = r.next
        return false if !r
        r = r.next
        
        return true if w == r
    end
    
    return false
end

File: 141-linked-list-cycle/141-linked-list-cycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(!head) return false;
        if(head->next == nullptr) return false;
        
        ListNode *turtle = head;
        ListNode *hare = head->next;
        
        for(int i=0; i<10000; ++i) {
            turtle = turtle->next;
            if(!hare) return false;
            hare = hare->next;
            if(!hare) return false;
            hare = hare->next;
            if(turtle == hare) return true;
        }
        
        return false;
    }
};